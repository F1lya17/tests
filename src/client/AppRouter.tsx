import React from "react";
import { Route, Switch } from "react-router";
import { Home } from "./pages/Home";
import { Catalog } from "./pages/Catalog";
import { Product } from "./pages/Product";
import { Delivery } from "./pages/Delivery";
import { Contacts } from "./pages/Contacts";
import { Cart } from "./pages/Cart";

const AppRouter = function () {
    return (
        <Switch>
            <Route path="/" exact component={Home} />
            <Route path="/catalog" exact component={Catalog} />
            <Route path="/catalog/:id" component={Product} />
            <Route path="/delivery" component={Delivery} />
            <Route path="/contacts" component={Contacts} />
            <Route path="/cart" component={Cart} />
        </Switch>
    );
}

export default AppRouter;