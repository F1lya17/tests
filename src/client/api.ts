import axios, { AxiosResponse } from 'axios';
import { CartState, CheckoutFormData, CheckoutResponse, Product, ProductShortInfo } from '../common/types';

export abstract class AbstractApi {
    constructor(protected readonly basename: string) {

    }

    abstract getProducts(): Promise<AxiosResponse<ProductShortInfo[]>>

    abstract getProductById(id: number): Promise<AxiosResponse<Product | null>>

    abstract checkout(form: CheckoutFormData, cart: CartState): Promise<AxiosResponse<CheckoutResponse>>
}

export class MockApi extends AbstractApi {
    constructor(protected readonly basename: string, private readonly products: ProductShortInfo[]) {
        super(basename);
        products = products;
    }

    getProducts(): Promise<AxiosResponse<ProductShortInfo[], any>> {
        return new Promise((resolve): void => {
            resolve({
                status: 200,
                statusText: "good",
                data: this.products,
                headers: {},
                config: {},
            })
        });
    }

    getProductById(id: number): Promise<AxiosResponse<Product, any>> {
        return new Promise((resolve) => {
        });
    }

    checkout(form: CheckoutFormData, cart: CartState): Promise<AxiosResponse<CheckoutResponse, any>> {
        return new Promise((resolve) => {
        });
    }
}

export class ExampleApi extends AbstractApi {
    constructor(protected readonly basename: string) {
        super(basename);
    }
    async getProducts() {
        let res = await axios.get<ProductShortInfo[]>(`${this.basename}/api/products`);
        return res;
    }

    async getProductById(id: number) {
        return await axios.get<Product>(`${this.basename}/api/products/${id}`);
    }

    async checkout(form: CheckoutFormData, cart: CartState) {
        return await axios.post<CheckoutResponse>(`${this.basename}/api/checkout`, { form, cart });
    }
}

export const LOCAL_STORAGE_CART_KEY = 'example-store-cart';

export abstract class CartAbstractApi {
    constructor(protected readonly cart: CartState) {
        this.cart = cart
    }
    abstract getState(): CartState

    abstract setState(cart: CartState): void
}

export class CartMockApi extends CartAbstractApi {
    getState() {
        return this.cart;
    }
    setState(cart: CartState) { }
}

export class CartApi extends CartAbstractApi {
    getState(): CartState {
        try {
            const json = localStorage.getItem(LOCAL_STORAGE_CART_KEY);
            return JSON.parse(json) as CartState || {};
        } catch {
            return {};
        }
    }

    setState(cart: CartState) {
        localStorage.setItem(LOCAL_STORAGE_CART_KEY, JSON.stringify(cart));
    }
}