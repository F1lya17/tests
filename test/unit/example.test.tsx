import React from 'react';
import TestComponent from '../TestComponent';
import axios from 'axios';
import { Provider } from 'react-redux';

import { ExampleApi, CartApi, MockApi, CartMockApi } from '../../src/client/api';
import { initStore } from '../../src/client/store';

import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom'
import { Catalog } from '../../src/client/pages/Catalog';
import { BrowserRouter } from 'react-router-dom';

const basename = '/';

describe('Общие требования', () => {
    it('при нажатии на название переходим на главную страницу', () => {
        const application = (
            <TestComponent />
        )
        render(application);
        const name = screen.getByTestId("title");
        fireEvent.click(name);
        expect(screen.getByTestId('home-page')).toBeInTheDocument();
    });

    it('в шапке отображаются ссылки на страницы магазина', () => {
        const application = (
            <TestComponent />
        )
        render(application);
        const catalog = screen.getByTestId("catalog").getAttribute("href");
        const delivery = screen.getByTestId("delivery").getAttribute("href");
        const contacts = screen.getByTestId("contacts").getAttribute("href");
        const cart = screen.getByTestId("cart").getAttribute("href");
        expect(catalog).toContain("/catalog");
        expect(delivery).toContain("/delivery");
        expect(contacts).toContain("/contacts");
        expect(cart).toContain("/cart");
    });
});


describe('Страницы', () => {
    it('в магазине должны быть страницы: главная, каталог, условия доставки, контакты', () => {
        const application = (
            <TestComponent />
        )
        render(application);
        const name = screen.getByTestId("title");
        fireEvent.click(name);
        expect(screen.getByTestId('home-page')).toBeInTheDocument();
        const catalog = screen.getByTestId("catalog");
        fireEvent.click(catalog);
        expect(screen.getByTestId('catalog-page')).toBeInTheDocument();
        const delivery = screen.getByTestId("delivery");
        fireEvent.click(delivery);
        expect(screen.getByTestId('delivery-page')).toBeInTheDocument();
        const contacts = screen.getByTestId("contacts");
        fireEvent.click(contacts);
        expect(screen.getByTestId('contacts-page')).toBeInTheDocument();
    })
})


describe('Каталог', () => {
    it('в каталоге должны отображаться товары, список которых приходит с сервера', async () => {
        const api = new MockApi(basename, [{ id: 777, name: "Anton", price: 100 }]);
        const cart = new CartMockApi({});
        const store = initStore(api, cart);
        render(
            <BrowserRouter basename=''>
                <Provider store={store}>
                    <Catalog />
                </Provider>
            </BrowserRouter>
        );
        await waitFor(() => screen.getByText('Anton'))
    })

    it('для каждого товара в каталоге отображается название, цена и ссылка на страницу с подробной информацией о товаре', async () => {
        const api = new MockApi(basename, [{ id: 777, name: "Anton", price: 100 }]);
        const cart = new CartMockApi({});
        const store = initStore(api, cart);
        render(
            <BrowserRouter basename=''>
                <Provider store={store}>
                    <Catalog />
                </Provider>
            </BrowserRouter>
        );
        await waitFor(() => expect(screen.getAllByRole('card')).toMatchSnapshot())
    })
})