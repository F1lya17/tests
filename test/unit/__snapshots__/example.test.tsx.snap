// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Каталог для каждого товара в каталоге отображается название, цена и ссылка на страницу с подробной информацией о товаре 1`] = `
[
  <div
    class="ProductItem card w-100 mb-4"
    data-testid="777"
    role="card"
  >
    <img
      class="Image card-img-top"
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkMAYAADkANVKH3ScAAAAASUVORK5CYII="
    />
    <div
      class="card-body"
    >
      <h5
        class="ProductItem-Name card-title"
      >
        Anton
      </h5>
      <p
        class="ProductItem-Price card-text"
      >
        $
        100
      </p>
      <a
        class="ProductItem-DetailsLink card-link"
        data-testid="link"
        href="/catalog/777"
      >
        Details
      </a>
    </div>
  </div>,
]
`;
